### DISCUSSION TITLE
post_232731

### DISCUSSION THREAD

**Discussion by cincuranet** [cincuranet/blog-comments#1187] [Comments]
imported 

https://www.tabsoverspaces.com/id/232731

**Comment by cincuranet** [comment]
Imported comment written by _Stephen Roughley_ on _2012-03-15T17:48:23_:

<p>Hi,</p><p>I had the same problem with needing to map to protected properties and your post led me to the following solution:</p><p>public static Expression<Func> GetFieldAccessor(string propertyName)<br>{<br>var param = Expression.Parameter(typeof(TEntity), "p");<br>var member = Expression.PropertyOrField(param, propertyName);</p><p>var expression = Expression.Lambda<Func>(member, param);<br>return expression;<br>}</p><p>Which you can call in your mappings like so:</p><p>modelBuilder.Entity().Property(ExpressionHelper.GetFieldAccessor("PropertyName")).HasColumnName("ColumnName");</p><p>Cheers,</p><p>Stephen</p>

**Comment by cincuranet** [comment]
Imported comment written by _Stephen Roughley_ on _2012-03-15T17:52:43_:

<p>Hhmmm... the formatting got kinda messed up there:</p><p>public static Expression<Func<TEntity, TResult>> GetFieldAccessor<TEntity, TResult>(string propertyName)<br>{<br>var param = Expression.Parameter(typeof(TEntity), "p");<br>var member = Expression.PropertyOrField(param, propertyName);</p><p>var expression = Expression.Lambda<Func<TEntity, TResult>>(member, param);<br>return expression;<br>}</p><p>With:</p><p>modelBuilder.Entity<Entity>().Property(ExpressionHelper.GetFieldAccessor<Entity, string>("PropertyName")).HasColumnName("ColumnName");</p><p>Hope this one works out better</p><p>Stephen</p>

**Comment by cincuranet** [comment]
Imported comment written by _Thiago Silva_ on _2013-05-24T00:25:31_:

<p>doesn't seem to work on properties that are collections...?</p>

**Comment by cincuranet** [comment]
Imported comment written by _Jiří Činčura ↹_ on _2013-05-24T00:35:21_:

<p>Why not? You just need to create expression and pass it to HasMany/WithMany/…, because such properties are basically associations.</p>

**Comment by cincuranet** [comment]
Imported comment written by _Thiago Silva_ on _2013-05-24T03:11:14_:

<p>ah...I see.  How about when using these properties in a LINQ query that needs an Include parameter (eager loading)?</p><p>e.g.:  _db.Customer.Include(p => p.MyProtectedProperty);</p><p>SInce the protected property is not accessible, what should be used in its place for the Include()?</p>

**Comment by cincuranet** [comment]
Imported comment written by _Jiří Činčura ↹_ on _2013-05-24T12:45:38_:

<p>Well it's non-public for a reason. If it's not accessible to you, maybe you should design your code different way. Or make it public. ;) Anyway, you can put expression into Include as well. BTW it takes also string, that might be easier too.</p>

**Comment by M-Tamir** [comment]
i get ExpressionHelper does not contain GetFieldAccessor.
I'm using EntityFramework" Version="6.4.4" , 
Microsoft.AspNet.Mvc" Version="5.2.7", 
Microsoft.EntityFrameworkCore" Version="5.0.11"

**Comment by cincuranet** [comment]
You need to use also the class in the comment.

Also looks weird to use both EF and EF Core at the same time.

**Comment by M-Tamir** [comment]
i was trying to use conventions to solve mapping a protected id using modelBuilder.Conventions and the EF got imported by VS .

yes I've used the class :
public static Expression GetFieldAccessor(int Id)
            {
                var param = Expression.Parameter(typeof(BaseAggregate), "p");
                var member = Expression.PropertyOrField(param, Id.ToString());
                var expression = Expression.Lambda(member, param);
                return expression;
            }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {       
 modelBuilder.Entity<BaseAggregate>().Property(ExpressionHelper.GetFieldAccessor("Id")).HasColumnName("Id");
}

**Comment by M-Tamir** [comment]
i think this fixed it : 
modelBuilder.Entity<BaseAggregate>().Property(typeof(int), "Id");
